civis/io/_tables.py

```diff
@@ -1310,168 +1292,144 @@ def _run_cleaning(file_ids, client, need_table_columns, headers, delimiter,
-    output_obj_id: int
-      The file ID under consideration; used for error messaging.
+    # The user cannot specify "compression" from civis_file_to_table.
+    if attr == "compression" or value_from_user is None:
+        return value_detected
 
-    Raises
-    ------
+    if value_detected != value_from_user:
+        raise CivisImportError(
+            f"All detected values for '{attr}' in your files "
+            f"are {value_detected}, which doesn't match "
+            f"your provided value of {value_from_user}"
+        )
+    else:
+        return value_detected
 
-    CivisImportError
-      If the table columns and the file columns have a type mismatch, or
-      differ in count.
-    """
-    if len(table_columns) != len(file_columns):
-        raise CivisImportError('All files should have the same number of '
-                               'columns. Expected {} columns but file {} '
-                               'has {} columns'.format(
-                                   len(table_columns),
-                                   output_obj_id,
-                                   len(file_columns))
-                               )
-
-    error_msgs = []
-    for idx, (tcol, fcol) in enumerate(zip(table_columns, file_columns)):
-        # for the purposes of type checking, we care only that the types
-        # share a base type (e.g. INT, VARCHAR, DECIMAl) rather than that
-        # they have the same precision and length
-        # (e.g VARCHAR(42), DECIMAL(8, 10))
-        tcol_base_type = tcol['sql_type'].split('(', 1)[0]
-        fcol_base_type = fcol['sql_type'].split('(', 1)[0]
-
-        if tcol_base_type != fcol_base_type:
-            error_msgs.append(
-                'Column {}: File base type was {}, but expected {}'.format(
-                    idx,
-                    fcol_base_type,
-                    tcol_base_type
-                )
-            )
-    if error_msgs:
+
+def _check_column_types(files: List[_File]):
+    cols_by_file: List[List[Dict[str, str]]]
+    cols_by_file = [f.detected_info["tableColumns"] for f in files]
+
+    col_counts = [len(cols) for cols in cols_by_file]
+    err_msg = _err_msg_if_inconsistent(col_counts, files)
+    if err_msg:
         raise CivisImportError(
-            'Encountered the following errors for file {}:\n\t{}'.format(
-                output_obj_id,
-                '\n\t'.join(error_msgs)
-            )
+            f"All files must have the same number of columns, "
+            f"however --\n{err_msg}"
         )
+
+    # Transpose cols_by_file to get cols_by_col
+    # https://stackoverflow.com/q/6473679
+    cols_by_col: List[List[Dict[str, str]]]
+    cols_by_col = list(map(list, zip(*cols_by_file)))
+
+    table_columns: List[Dict[str, str]] = []
+    allow_inconsistent_headers = False
+    err_msgs: List[str] = []
+
+    for i, cols in enumerate(cols_by_col, 1):
+        col_name = cols[0].get("name") or f"column_{i}"
```

Comment:
I found an issue where this didn't take the correct column name when the first file happened to be the one with only strings.  This code is a little ugly, but will take the first truthy value or default to the `column_{i}` form if there are no truthy values.

```suggestion
        col_name = next((c.get("name") for c in cols if c.get("name")), f"column_{i}")
```